src/hud.cpp

```diff
@@ -489,23 +491,96 @@ void drawItemStack(video::IVideoDriver *driver,
+		bool hovered,
+		bool dragged)
 {
-	if(item.empty())
+	static s32 hovered_time;
+	static s32 selected_time;
+	static s32 dragged_time;
+	static scene::IMesh *hovered_mesh;
+	static scene::IMesh *selected_mesh;
+	static scene::IMesh *dragged_mesh;
+	bool enable_animations = 
+		g_settings->getBool("enable_inventory_animations");	
+
+	if(item.empty()) {
+		if (selected) {
+			selected_mesh = NULL;
+		} else if (hovered) {
+			hovered_mesh = NULL;
+		} else if (dragged) {
+			dragged_mesh = NULL;
+		}
 		return;
+	}
 
 	const ItemDefinition &def = item.getDefinition(gamedef->idef());
-	video::ITexture *texture = gamedef->idef()->getInventoryTexture(def.name, gamedef);
+	scene::IMesh* mesh = gamedef->idef()->getWieldMesh(def.name, gamedef);
+
+	if (mesh) {
+		s32 delta = 0;
+		if (selected) {
+			if (mesh != selected_mesh) {
+				selected_mesh = mesh;
+				selected_time = getTimeMs();
+			}
+			delta = porting::getDeltaMs(selected_time, getTimeMs()) % 100000;
+		} else if (hovered) {
+			if (mesh != hovered_mesh) {
+				hovered_mesh = mesh;
+				hovered_time = getTimeMs();
+			}
+			delta = porting::getDeltaMs(hovered_time, getTimeMs()) % 100000;
+		} else if (dragged) {
+			if (mesh != dragged_mesh) {
+				dragged_mesh = mesh;
+				dragged_time = getTimeMs();
+			}
+			delta = porting::getDeltaMs(dragged_time, getTimeMs()) % 100000;
+		}
+		core::rect<s32> oldViewPort = driver->getViewPort();
+		core::matrix4 oldProjMat = driver->getTransform(video::ETS_PROJECTION);
+		core::matrix4 oldViewMat = driver->getTransform(video::ETS_VIEW);
+		core::matrix4 ProjMatrix;
+		ProjMatrix.buildProjectionMatrixOrthoLH(2, 2, -1, 100); 
+		driver->setTransform(video::ETS_PROJECTION, ProjMatrix);
+		driver->setTransform(video::ETS_VIEW, ProjMatrix);
+		core::matrix4 matrix;
+		matrix.makeIdentity();
+
+		if (enable_animations) {
+			float timer_f = (float)delta / 5000.0;
+			matrix.setRotationDegrees(core::vector3df(0, 360 * timer_f, 0));
+		}
 
-	// Draw the inventory texture
-	if(texture != NULL)
-	{
-		const video::SColor color(255,255,255,255);
-		const video::SColor colors[] = {color,color,color,color};
-		draw2DImageFilterScaled(driver, texture, rect,
-			core::rect<s32>(core::position2d<s32>(0,0),
-			core::dimension2di(texture->getOriginalSize())),
-			clip, colors, true);
+		driver->setTransform(video::ETS_WORLD, matrix);
+		driver->setViewPort(rect);
+
+		u32 mc = mesh->getMeshBufferCount();
+		for (u32 j = 0; j < mc; ++j) {
+			scene::IMeshBuffer *buf = mesh->getMeshBuffer(j);
+			video::SMaterial &material = buf->getMaterial();
+			material.MaterialType = video::EMT_TRANSPARENT_ALPHA_CHANNEL_REF;
+			material.Lighting = false;
+			driver->setMaterial(material);
+			driver->drawMeshBuffer(buf);
+		}
+
+		driver->setTransform(video::ETS_VIEW, oldViewMat);
+		driver->setTransform(video::ETS_PROJECTION, oldProjMat);
+		driver->setViewPort(oldViewPort);
+	} else {
+		video::ITexture *texture = gamedef->idef()->getInventoryTexture(def.name, gamedef);
```

Comment:
in fact i can trash that code, else wont never ever be executed, theres no way for mesh to be unaviable